From 2b73d00c85edb8653e063975346a4ac4a532ac5a Mon Sep 17 00:00:00 2001
From: ruslanbay <67730802+ruslanbay@users.noreply.github.com>
Date: Wed, 10 Dec 2025 15:09:22 +0100
Subject: [PATCH 17/29] add .vidioc_enum_framesizes =
 ipu_isys_vidioc_enum_framesizes to resolve `Unable to enumerate frame sizes:
 Inappropriate ioctl for device`:

```
[0:03:44.690752721] [3423] DEBUG SimplePipeline simple.cpp:861 Link 'ov8865 1-0010'[0] -> 'Intel IPU4 CSI-2 3'[0]: configured with format 3264x2448-SBGGR10_1X10/RAW
[0:03:44.690779220] [3423] DEBUG SimplePipeline simple.cpp:861 Link 'Intel IPU4 CSI-2 3'[1] -> 'Intel IPU4 CSI-2 3 capture 0'[0]: configured with format 3264x2448-SBGGR10_1X10/Unset
[0:03:44.691131195] [3423] ERROR V4L2 v4l2_videodevice.cpp:1233 /dev/video15[13:cap]: Unable to enumerate frame sizes: Inappropriate ioctl for device
[0:03:44.691171006] [3423] DEBUG SimplePipeline simple.cpp:707 Adding configuration for 3264x2448 in pixel formats [  ]
[0:03:44.691202612] [3423] ERROR SimplePipeline simple.cpp:644 No valid configuration found
```
---
 drivers/media/pci/intel/ipu-isys-video.c | 27 ++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/drivers/media/pci/intel/ipu-isys-video.c b/drivers/media/pci/intel/ipu-isys-video.c
index 5a8c8d22f877..ede2f541cf7f 100644
--- a/drivers/media/pci/intel/ipu-isys-video.c
+++ b/drivers/media/pci/intel/ipu-isys-video.c
@@ -459,6 +459,32 @@ int ipu_isys_vidioc_enum_fmt(struct file *file, void *fh,
 	return -EINVAL;
 }
 
+static int ipu_isys_vidioc_enum_framesizes(struct file *file, void *fh,
+					    struct v4l2_frmsizeenum *fsize)
+{
+	unsigned int i;
+
+	if (fsize->index > 0)
+		return -EINVAL;
+
+	for (i = 0; i < ARRAY_SIZE(ipu_isys_pfmts); i++) {
+		if (fsize->pixel_format != ipu_isys_pfmts[i].pixelformat)
+			continue;
+
+		fsize->type = V4L2_FRMSIZE_TYPE_STEPWISE;
+		fsize->stepwise.min_width = IPU_ISYS_MIN_WIDTH;
+		fsize->stepwise.max_width = IPU_ISYS_MAX_WIDTH;
+		fsize->stepwise.min_height = IPU_ISYS_MIN_HEIGHT;
+		fsize->stepwise.max_height = IPU_ISYS_MAX_HEIGHT;
+		fsize->stepwise.step_width = 2;
+		fsize->stepwise.step_height = 2;
+
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
 static int vidioc_g_fmt_vid_cap_mplane(struct file *file, void *fh,
 				       struct v4l2_format *fmt)
 {
@@ -2099,6 +2125,7 @@ static const struct v4l2_ioctl_ops ioctl_ops_splane = {
 static const struct v4l2_ioctl_ops ioctl_ops_mplane = {
 	.vidioc_querycap = ipu_isys_vidioc_querycap,
 	.vidioc_enum_fmt_vid_cap = ipu_isys_vidioc_enum_fmt,
+	.vidioc_enum_framesizes = ipu_isys_vidioc_enum_framesizes,
 	.vidioc_g_fmt_vid_cap_mplane = vidioc_g_fmt_vid_cap_mplane,
 	.vidioc_s_fmt_vid_cap_mplane = vidioc_s_fmt_vid_cap_mplane,
 	.vidioc_try_fmt_vid_cap_mplane = vidioc_try_fmt_vid_cap_mplane,
-- 
2.51.0

